File: aoj/src/GRL_5_A/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

var sc = bufio.NewScanner(os.Stdin)
var adj [][]edge

func dfs(current int, weight int, visited []bool) (int, int) {
	max := weight
	dst := current
	visited[current] = true

	for _, e := range adj[current] {
		if visited[e.dst] {
			continue
		}
		w, d := dfs(e.dst, weight+int(e.weight), visited)
		if w > max {
			max = w
			dst = d
		}
	}
	return max, dst
}

type edge struct {
	dst    int
	weight int16
}

//  GRL_5_A: 木の直径
func main() {
	sc.Buffer([]byte{}, math.MaxInt64)
	sc.Split(bufio.ScanWords)
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()

	n := nextInt()

	adj = make([][]edge, n, n)
	for i := 0; i < n; i++ {
		adj[i] = make([]edge, 0, 0)
	}
	for i := 0; i < n; i++ {
		src := nextInt()
		dst := nextInt()
		weight := int16(nextInt())
		adj[src] = append(adj[src], edge{
			dst:    dst,
			weight: weight,
		})
		adj[dst] = append(adj[dst], edge{
			dst:    src,
			weight: weight,
		})

	}
	_, d := dfs(0, 0, make([]bool, n, n))
	weight, _ := dfs(d, 0, make([]bool, n, n))
	fmt.Fprintln(w, weight)
}

func next() string {
	sc.Scan()
	return sc.Text()
}

func nextInt() int {
	s := next()
	i, _ := strconv.Atoi(s)
	return i
}

File: aoj/src/GRL_5_A/main.py
import sys

class resolver:
    def __init__(self, adj):
        self.adj = adj

    def farthest(self, i):
        next = [(i, 0)]
        visited = set([i])
        at = i
        aw = 0
        while(len(next) != 0):
            i, tw = next.pop(0)
            for t, pw in self.adj[i]:
                if not t in visited:
                    nw = tw +pw
                    visited.add(t)
                    next.append((t, nw))
                    if aw < nw:
                        aw = nw
                        at = t
        return at, aw

    def resolve(self):
        s, _ = self.farthest(0)
        _, w = self.farthest(s)
        print(w)

# GRL_5_A: 木の直径
def main():
    readline = sys.stdin.readline
    n = int(input())
    adj = [[] for _ in range(n)]
    for _ in range(n-1):
        s, t, w = map(int, readline().split())
        adj[s].append((t, w))
        adj[t].append((s, w))
    resolver(adj).resolve()
    
if __name__ == "__main__":
    main()
